Source/JavaScriptCore/runtime/JSArray.cpp

```diff
@@ -481,6 +481,65 @@ bool JSArray::setLengthWithArrayStorage(JSGlobalObject* globalObject, unsigned n
     return true;
 }
 
+bool JSArray::fastFill(VM& vm, unsigned startIndex, unsigned endIndex, JSValue value)
+{
+    if (isCopyOnWrite(indexingMode()))
+        convertFromCopyOnWrite(vm);
+
+    IndexingType type = indexingType();
+
+    auto getNextIndexingType = [type, value]() {
+        if (!(type & IsArray))
+            return NonArray;
+        if (hasAnyArrayStorage(type))
+            return NonArray;
+        switch (type) {
+        case ArrayWithInt32:
+        case ArrayWithUndecided:
+            if (value.isInt32())
+                return ArrayWithInt32;
+            if (value.isNumber())
+                return ArrayWithDouble;
+            return ArrayWithContiguous;
+        case ArrayWithDouble:
+            if (value.isNumber())
+                return type;
+            return ArrayWithContiguous;
+        case ArrayWithContiguous:
+            return type;
+        default:
+            return NonArray;
+        }
+    };
+    IndexingType nextType = getNextIndexingType();
+    if (type == ArrayWithUndecided) {
+        if (nextType == ArrayWithInt32)
+            convertUndecidedToInt32(vm);
+        else if (nextType == ArrayWithDouble)
+            convertUndecidedToDouble(vm);
+        else if (nextType == ArrayWithContiguous)
+            convertUndecidedToContiguous(vm);
+        else {
+            ASSERT_NOT_REACHED();
+            return false;
+        }
+    } else if (type != nextType)
+        return false;
+
+    ASSERT(nextType == indexingType());
+
+    if (nextType == ArrayWithDouble) {
+        auto* data = butterfly()->contiguousDouble().data();
+        std::fill(data + startIndex, data + endIndex, value.asNumber());
+    } else {
+        auto* data = butterfly()->contiguous().data();
+        std::fill(data + startIndex, data + endIndex, std::bit_cast<const WriteBarrier<Unknown>>(JSValue::encode(value)));
```

Comment:
For `ArrayWithInt32` case, this is fine, but for ArrayWithContiguous case, this is not correct since concurrent GC can scan this butterfly while writing, and if the written value is not fully commited (like, byte-by-byte write), then it becomes broken cell in the concurrent GC marker.
You need to do something like `gcSafeMemcpy` but for memset.

Reply:
For MacOS, can we use `memset_pattern8` here?

Reply:
Implementing a GC safe memset is difficult for me, so I used `memset_pattern8` on MacOS and a for loop for other OS.